### PR TITLE
Upgrade rawdb and statedb codes to add the latest functionalities of ethdb

### PR DESCRIPTION
**Pull request by GheisMohammadi**
## Issue

The **RawDB** and **StateDB** in core are using Ethereum codes. These have been in place for a while, and since then, the ethdb has received a couple of extra functions and improvements. One of the features we need for implementing state pruning is snapshot. To be able to add snapshot, the rawdb and statedb should be updated with newest version of codes. This PR burrows those codes and adds them to core. 

## Test

run a node and monitor and check how it performs.

## Operational Checklist

1. **Does this PR introduce backward-incompatible changes to the on-disk data structure and/or the over-the-wire protocol?**. (If no, skip to question 8.)
    **NO**

2. **Does this PR introduce backward-incompatible changes *NOT* related to on-disk data structure and/or over-the-wire protocol?** (If no, continue to question 11.)
    **NO**

3. **Does the existing `node.sh` continue to work with this change?**  
    **YES**

4. **What should node operators know about this change?**  
    It's backward compatible and no need to do anything

### CODE REVIEW

**Comment by ONECasey** [comment]
Conflicts fixed. Waiting for tests before merge.

**Review by ONECasey — APPROVED** [review]
Looks good. This is a huge PR so it took a little while to review and approve.

**Comment by GheisMohammadi** [comment]
> Conflicts fixed. Waiting for tests before merge.

Great, thanks, tests are in progress. We are testing on a few validator/explorer nodes in testnet and mainnet.

File: core/rawdb/accessors_snapshot.go
Comment (ONECasey) [reply core/rawdb/accessors_snapshot.go:35]: This is a play of The Hitchhiker's Guide to the Galaxy. The value 42 is used as a placeholder or arbitrary value to indicate that the "snapshot pause flag" has been disabled. It is not a special value in particular but rather a number that has been used humorously

Comment (MaxMustermann2) [reply core/rawdb/accessors_snapshot.go:35]: Thanks, I am aware of the reference. But good to ask for clarification.

File: core/rawdb/accessors_chain.go
Comment (GheisMohammadi) [reply core/rawdb/accessors_chain.go:640]: It is just ported over. Ethereum uses this to store the txlookup limit in db. It is used in Ethereum fast sync. We can remove it for now.

File: core/rawdb/accessors_chain.go
Comment (GheisMohammadi) [reply core/rawdb/accessors_chain.go:802]: Same, just ported over. We don't need it yet. We can remove it for now.

**Pull request merged**
